instrumentation/github.com/astaxie/beego/beego.go

```diff
@@ -0,0 +1,127 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package beego
+
+import (
+	"context"
+	"net/http"
+
+	"google.golang.org/grpc/codes"
+
+	otelhttp "go.opentelemetry.io/contrib/instrumentation/net/http"
+	"go.opentelemetry.io/otel/api/trace"
+
+	"github.com/astaxie/beego"
+)
+
+// defaultSpanNameFormatter is the default formatter for spans created with
+// the beego integration. Returns the request URL path.
+func defaultSpanNameFormatter(operation string, req *http.Request) string {
+	if req.URL.Path == "" {
+		return req.Method
+	}
+	return req.URL.Path
```

Comment:
Is there any way this can be made lower cardinality by using the route template instead of all the actual paths?

Reply:
Okay so I've been working on this and I think I've come up with a solution. I would love to get your feedback on it, because its a little tricky. You can see what the diff looks like [here](https://github.com/reggiemcdonald/opentelemetry-go-contrib/compare/82-beego-integ...reggiemcdonald:82-beego-integ-span-name)

The problem is that the function beego provides to get the route template from a URL requires a `github.com/astaxie/beego/context.Contex`. You can get one of these contexts, but you need to set them with a `http.ResponseWriter` and `*http.Request` ([link to code](https://github.com/reggiemcdonald/opentelemetry-go-contrib/blob/950f7b2f90f670f244cb091071fcd48617d99ac8/instrumentation/github.com/astaxie/beego/beego.go#L36-L38)).

In the `SpanNameFormatter`, there is no `ResponseWriter` available to me. So I can't properly search up the route template from the `SpanNameFormatter`. 

So to solve this dilemma, what I've done is wrap the `ServeHTTP` function to lookup the route template and save it to the context before calling `otelhttp.ServeHTTP`. Then the `SpanNameFormatter` can retrieve the route template from the context if it exists, and return it as the span name. If the value isn't in the context (serving static files or using a router outside of the `beego.BeeApp`), then the HTTP method name is returned. In either case, cardinality is low.

Reply:
This looks like a decent approach to me. :+1: 

I would say it would also be worth opening an issue in github.com/astaxie/beego to see if they can wrap this functionality themselves or maybe provide the template path in a more direct way to the `SpanNameFormatter`. Maybe by adding this to their context implementation? LMK if you would like help opening the issue.

Reply:
Awesome! I'll open an issue asking them to consider providing a function that can get the route template with just an http.Request, since looking through the code, thats all thats actually being used

Reply:
[#4162](https://github.com/astaxie/beego/issues/4162)

Reply:
The beego owners got back to me. An interceptor API will be released sometime soon that would make getting the template path more direct. They recently merged an example of using the new interface for opentracing [here](https://github.com/astaxie/beego/pull/4130).

Downside of the new approach is that it wouldn't let us use the existing net/http instrumentation package.

Did you want me to close this PR and start working on an implementation that uses the interceptor?

Reply:
It looks like the `FilterChain` implementation described there is not yet available in a released version of beego.  Either approach is probably fine, but to rewrite using `FilterChain` we'd need to hold off shipping this instrumentation until after the next beego release.

Reply:
> Either approach is probably fine, but to rewrite using FilterChain we'd need to hold off shipping this instrumentation until after the next beego release.

Let's go with the current approach for now and we can file a ticket that will track this as a future to possibly restructure. Thoughts?

Reply:
Yea I think thats fair. It doesn't seem like the `FilterChain` provides any functionality above the current implementation aside from making it easier to get the beego context

Reply:
#234